File: sdk/packages/python/iii/src/iii/stream.py
"""Stream types and interfaces for the III SDK."""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Any, Generic, List, Literal, TypeVar

from pydantic import BaseModel, Field, model_serializer

TData = TypeVar("TData")


class StreamAuthInput(BaseModel):
    """Input for stream authentication."""

    headers: dict[str, str]
    path: str
    query_params: dict[str, list[str]]
    addr: str


class StreamAuthResult(BaseModel):
    """Result of stream authentication."""

    context: Any | None = None


StreamContext = Any


class StreamJoinLeaveEvent(BaseModel):
    """Event for stream join/leave."""

    subscription_id: str
    stream_name: str
    group_id: str
    id: str | None = None
    context: Any | None = None


class StreamJoinResult(BaseModel):
    """Result of stream join."""

    unauthorized: bool


class StreamGetInput(BaseModel):
    """Input for stream get operation."""

    stream_name: str
    group_id: str
    item_id: str


class StreamSetInput(BaseModel):
    """Input for stream set operation."""

    stream_name: str
    group_id: str
    item_id: str
    data: Any


class StreamDeleteInput(BaseModel):
    """Input for stream delete operation."""

    stream_name: str
    group_id: str
    item_id: str


class StreamListInput(BaseModel):
    """Input for stream list operation."""

    stream_name: str
    group_id: str


class StreamListGroupsInput(BaseModel):
    """Input for stream list groups operation."""

    stream_name: str


class StreamUpdateInput(BaseModel):
    """Input for stream update operation."""

    stream_name: str
    group_id: str
    item_id: str
    ops: list["UpdateOp"]


class UpdateOpError(BaseModel):
    """Per-op error returned by ``state::update`` / ``stream::update``.

    Currently emitted only by the ``merge`` op when input violates the
    new validation bounds. Successfully applied ops are still
    reflected in the response's ``new_value``.
    """

    op_index: int
    code: str
    message: str
    doc_url: str | None = None


class StreamSetResult(BaseModel, Generic[TData]):
    """Result of stream set operation."""

    old_value: TData | None = None
    new_value: TData


class StreamUpdateResult(BaseModel, Generic[TData]):
    """Result of stream update operation."""

    old_value: TData | None = None
    new_value: TData
    # Per-op errors. Emitted by ``merge`` and ``append`` for validation
    # rejections (path/value bounds, proto-pollution segments) and by
    # ``append`` for the ``append.type_mismatch`` and
    # ``append.target_not_object`` surfaces. Field is omitted from the
    # JSON wire when empty. ``default_factory`` is used (not ``= []``)
    # to keep Pydantic's parameterized-Generic + default handling
    # well-behaved across Python versions.
    errors: list[UpdateOpError] = Field(default_factory=list)


class StreamDeleteResult(BaseModel):
    """Result of stream delete operation."""

    old_value: Any | None = None


class UpdateSet(BaseModel):
    """Set operation for stream update."""

    type: str = "set"
    path: str
    value: Any


class UpdateIncrement(BaseModel):
    """Increment operation for stream update."""

    type: str = "increment"
    path: str
    by: int | float


class UpdateDecrement(BaseModel):
    """Decrement operation for stream update."""

    type: str = "decrement"
    path: str
    by: int | float


class UpdateAppend(BaseModel):
    """Append an element to an array, concatenate a string, or push at a nested path.

    The target is the root (when ``path`` is omitted, an empty string,
    or an empty list), a single first-level key (when ``path`` is a
    non-empty string), or an arbitrary nested location (when ``path``
    is a list of literal segments).

    Path forms accepted (mirrors :class:`UpdateMerge` after #1547):
      - ``None`` / ``""`` / ``[]``: append at the root.
      - ``"foo"``: append at the first-level key ``foo``. A dotted
        string like ``"a.b"`` is the literal key ``"a.b"``, *not*
        traversed as ``a -> b``.
      - ``["a", "b", "c"]``: nested path; each element is a literal
        segment.

    Engine semantics:
      - Missing/non-object intermediates along a nested path are
        auto-created/replaced with ``{}``.
      - At the leaf:
          - missing/null + nested path -> ``[value]`` (always an array)
          - missing/null + single-string path -> string-as-string for
            the string-concat tier, otherwise ``[value]``
          - existing array -> push
          - existing string + string value -> concatenate
          - existing object/scalar at the leaf -> ``append.type_mismatch``

    Validation: invalid paths (depth > 32 segments, segment > 256
    bytes, or any ``__proto__`` / ``constructor`` / ``prototype``
    segment) are rejected with a structured error in the ``errors``
    field of the ``state::update`` / ``stream::update`` response. The
    append does not apply when an error is returned for that op.
    """

    type: str = "append"
    # Optional. Accepts a single string (legacy / first-level key) or
    # a list of literal segments (nested append). ``None`` / ``""`` /
    # ``[]`` all route to root append.
    path: str | list[str] | None = None
    value: Any

    @model_serializer(mode="wrap")
    def _omit_none_path(self, handler):  # type: ignore[no-untyped-def]
        # Drop ``path: None`` from the wire so cross-SDK consumers see
        # the field absent rather than ``null``. Mirrors the Rust
        # ``#[serde(skip_serializing_if = "Option::is_none")]`` on
        # ``UpdateOp::Append.path``.
        data = handler(self)
        if data.get("path") is None:
            data.pop("path", None)
        return data


class UpdateRemove(BaseModel):
    """Remove operation for stream update."""

    type: str = "remove"
    path: str


class UpdateMerge(BaseModel):
    """Shallow merge an object into the target.

    The target is the root (when ``path`` is omitted, an empty string,
    or an empty list) or an arbitrary nested location specified by an
    array of literal segments.

    Path forms accepted:
      - ``None`` / ``""`` / ``[]``: merge at the root.
      - ``"foo"``: equivalent to ``["foo"]`` -- single first-level key.
      - ``["a", "b", "c"]``: nested path. Each element is a *literal*
        key. ``["a.b"]`` writes a single key named ``"a.b"``, not
        ``a -> b``.

    Engine semantics:
      - Missing or non-object intermediates along the path are
        auto-replaced with ``{}``.
      - The merge is shallow at the target node (top-level keys of
        ``value`` overwrite same-named keys; siblings preserved).

    Validation: invalid paths/values (depth > 32 segments, segment >
    256 bytes, value depth > 16, > 1024 top-level keys, or any
    ``__proto__`` / ``constructor`` / ``prototype`` segment or
    top-level key) are rejected with a structured error in the
    ``errors`` array of the ``state::update`` / ``stream::update``
    response. The merge does not apply when an error is returned.
    """

    type: str = "merge"
    # Optional. Accepts a single string or a list of literal segments.
    # Pydantic resolves ``str | list[str]`` via smart-union: string
    # input -> str, array input -> list[str].
    path: str | list[str] | None = None
    value: Any

    @model_serializer(mode="wrap")
    def _omit_none_path(self, handler):  # type: ignore[no-untyped-def]
        # Mirrors the same skip-when-none rule applied to
        # ``UpdateOp::Merge.path`` in the Rust SDK so cross-SDK wire
        # payloads are byte-identical for root merges.
        data = handler(self)
        if data.get("path") is None:
            data.pop("path", None)
        return data


UpdateOp = UpdateSet | UpdateIncrement | UpdateDecrement | UpdateAppend | UpdateRemove | UpdateMerge


class StreamTriggerConfig(BaseModel):
    """Trigger config for ``stream`` triggers. Filters which item changes fire the handler."""

    stream_name: str
    group_id: str | None = None
    item_id: str | None = None
    condition_function_id: str | None = None


class StreamJoinLeaveTriggerConfig(BaseModel):
    """Trigger config for ``stream:join`` and ``stream:leave`` triggers."""

    condition_function_id: str | None = None


class StreamChangeEventDetail(BaseModel):
    """Detail of a stream change event containing the mutation type and data."""

    type: Literal["create", "update", "delete"]
    data: Any


class StreamChangeEvent(BaseModel):
    """Handler input for ``stream`` triggers, fired when an item changes."""

    type: Literal["stream"]
    timestamp: int
    streamName: str
    groupId: str
    id: str | None = None
    event: StreamChangeEventDetail


class IStream(ABC, Generic[TData]):
    """Abstract interface for stream operations."""

    @abstractmethod
    async def get(self, input: StreamGetInput) -> TData | None:
        """Get an item from the stream."""
        ...

    @abstractmethod
    async def set(self, input: StreamSetInput) -> StreamSetResult[TData] | None:
        """Set an item in the stream."""
        ...

    @abstractmethod
    async def delete(self, input: StreamDeleteInput) -> StreamDeleteResult:
        """Delete an item from the stream."""
        ...

    @abstractmethod
    async def list(self, input: StreamListInput) -> list[TData]:
        """Get all items in a group."""
        ...

    @abstractmethod
    async def list_groups(self, input: StreamListGroupsInput) -> List[str]:
        """List all groups in the stream."""
        ...

    @abstractmethod
    async def update(self, input: StreamUpdateInput) -> StreamUpdateResult[TData] | None:
        """Apply atomic update operations to a stream item."""
        ...

File: sdk/packages/node/iii/src/stream.ts
/** Input for stream authentication. */
export interface StreamAuthInput {
  /** Request headers. */
  headers: Record<string, string>
  /** Request path. */
  path: string
  /** Query parameters. */
  query_params: Record<string, string[]>
  /** Client address. */
  addr: string
}

/** Result of stream authentication. */
export interface StreamAuthResult {
  /** Arbitrary context passed to stream handlers after authentication. */
  // biome-ignore lint/suspicious/noExplicitAny: any is fine here
  context?: any
}

/** Context type extracted from {@link StreamAuthResult}. */
export type StreamContext = StreamAuthResult['context']

/** Event payload for stream join/leave events. */
export interface StreamJoinLeaveEvent {
  /** Unique subscription identifier. */
  subscription_id: string
  /** Name of the stream. */
  stream_name: string
  /** Group identifier. */
  group_id: string
  /** Item identifier (if applicable). */
  id?: string
  /** Auth context from {@link StreamAuthResult}. */
  context?: StreamContext
}

/** Result of a stream join request. */
export interface StreamJoinResult {
  /** Whether the join was unauthorized. */
  unauthorized: boolean
}

/** Input for retrieving a single stream item. */
export type StreamGetInput = {
  /** Name of the stream. */
  stream_name: string
  /** Group identifier. */
  group_id: string
  /** Item identifier. */
  item_id: string
}

/** Input for setting a stream item. */
export type StreamSetInput = {
  /** Name of the stream. */
  stream_name: string
  /** Group identifier. */
  group_id: string
  /** Item identifier. */
  item_id: string
  /** Data to store. */
  // biome-ignore lint/suspicious/noExplicitAny: any is fine here
  data: any
}

/** Input for deleting a stream item. */
export type StreamDeleteInput = {
  /** Name of the stream. */
  stream_name: string
  /** Group identifier. */
  group_id: string
  /** Item identifier. */
  item_id: string
}

/** Input for listing all items in a stream group. */
export type StreamListInput = {
  /** Name of the stream. */
  stream_name: string
  /** Group identifier. */
  group_id: string
}

/** Input for listing all groups in a stream. */
export type StreamListGroupsInput = {
  /** Name of the stream. */
  stream_name: string
}

/** Result of a stream set operation. */
export type StreamSetResult<TData> = {
  /** Previous value (if it existed). */
  old_value?: TData
  /** New value that was stored. */
  new_value: TData
}

/** Result of a stream update operation. */
export type StreamUpdateResult<TData> = {
  /** Previous value (if it existed). */
  old_value?: TData
  /** New value after the update. */
  new_value: TData
  /**
   * Per-op errors. Emitted by `merge` and `append` for validation
   * rejections (path depth/size, value depth, or a
   * `__proto__`/`constructor`/`prototype` segment or top-level key)
   * and by `append` for the case-2 `append.type_mismatch` and
   * `append.target_not_object` surfaces. Successfully applied ops are
   * still reflected in `new_value`. The field is omitted from the
   * JSON wire when empty.
   */
  errors?: UpdateOpError[]
}

/** Set a field at the given path to a value. */
export type UpdateSet = {
  type: 'set'
  /** First-level field path. Use an empty string to target the root value. */
  path: string
  /** Value to set. */
  // biome-ignore lint/suspicious/noExplicitAny: any is fine here
  value: any
}

/** Increment a numeric field by a given amount. */
export type UpdateIncrement = {
  type: 'increment'
  /** First-level field path. */
  path: string
  /** Amount to increment by. */
  by: number
}

/** Decrement a numeric field by a given amount. */
export type UpdateDecrement = {
  type: 'decrement'
  /** First-level field path. */
  path: string
  /** Amount to decrement by. */
  by: number
}

/**
 * Append an element to an array, concatenate a string, or push a new
 * value at a nested path. The target is the root (when `path` is
 * omitted, empty, or `[]`), a single first-level key (when `path` is
 * a non-empty string), or an arbitrary nested location (when `path`
 * is an array of literal segments).
 *
 * Engine semantics:
 *   - Missing or non-object intermediates along a nested path are
 *     auto-replaced with `{}` so a stray `null` or scalar never blocks
 *     future appends.
 *   - At the leaf:
 *       - missing/null + nested path → `[value]` (always an array)
 *       - missing/null + single-string path → string-as-string for the
 *         string-concat tier, otherwise `[value]`
 *       - existing array → push
 *       - existing string + string value → concatenate
 *       - existing object/scalar at the leaf → `append.type_mismatch`
 *   - Each path segment is a literal key. `["a.b"]` targets a single
 *     key named `"a.b"`, not `a → b`.
 *
 * Validation: invalid paths (depth > 32 segments, segment > 256
 * bytes, or any `__proto__`/`constructor`/`prototype` segment) are
 * rejected with a structured error in the `errors` field of the
 * `state::update` / `stream::update` response. The append does not
 * apply when an error is returned for that op.
 */
export type UpdateAppend = {
  type: 'append'
  /**
   * Optional path to the append target. Accepts a single first-level
   * key (legacy `string`) or an array of literal segments for nested
   * append. See {@link MergePath} (the same shape is reused).
   */
  path?: MergePath
  /** Value to append. String targets only accept string values. */
  // biome-ignore lint/suspicious/noExplicitAny: any is fine here
  value: any
}

/** Remove a field at the given path. */
export type UpdateRemove = {
  type: 'remove'
  /** First-level field path. */
  path: string
}

/**
 * Path target for a {@link UpdateMerge} op. Accepts:
 *   - a single string (legacy / first-level field)
 *   - an array of literal segments (nested path; each element is one
 *     literal key, dots are NOT interpreted as separators)
 *
 * Omit `path`, pass `""`, or pass `[]` to target the root value.
 *
 * @example single first-level field
 *   { type: 'merge', path: 'session-abc', value: { author: 'alice' } }
 * @example nested path (closes issue #1546's structured-state use case)
 *   { type: 'merge', path: ['sessions', 'abc'], value: { ts: chunk } }
 */
export type MergePath = string | string[]

/**
 * Shallow-merge an object into the target. The target is the root
 * (when `path` is omitted/empty) or an arbitrary nested location
 * specified by an array of literal segments.
 *
 * Engine semantics:
 *   - Missing or non-object intermediates along the path are
 *     auto-replaced with `{}` so a stray `null` or scalar never
 *     blocks future merges.
 *   - The merge is shallow at the target — top-level keys of `value`
 *     replace same-named keys; siblings are preserved.
 *   - Each path segment is a literal key. `["a.b"]` writes a single
 *     key named `"a.b"`, not `a → b`.
 *
 * Validation: invalid paths/values (depth > 32 segments, segment >
 * 256 bytes, value depth > 16, > 1024 top-level keys, or any
 * `__proto__`/`constructor`/`prototype` segment or top-level key)
 * are rejected with a structured error in the `errors` field of the
 * `state::update` / `stream::update` response. The merge does not
 * apply when an error is returned for that op.
 */
export type UpdateMerge = {
  type: 'merge'
  /** Optional path to the merge target. See {@link MergePath}. */
  path?: MergePath
  /** Object to merge. Must be a JSON object. */
  // biome-ignore lint/suspicious/noExplicitAny: any is fine here
  value: any
}

/** Per-op error returned by `state::update` / `stream::update`. */
export type UpdateOpError = {
  /** Index of the offending op within the original `ops` array. */
  op_index: number
  /** Stable error code, e.g. `"merge.path.too_deep"`. */
  code: string
  /** Human-readable description with concrete numbers when applicable. */
  message: string
  /** Optional documentation URL. */
  doc_url?: string
}

/** Result of a stream delete operation. */
export type DeleteResult = {
  /** Previous value (if it existed). */
  // biome-ignore lint/suspicious/noExplicitAny: any is fine here
  old_value?: any
}

/**
 * Union of all atomic update operations supported by streams.
 *
 * @see {@link UpdateSet}, {@link UpdateIncrement}, {@link UpdateDecrement},
 *      {@link UpdateAppend}, {@link UpdateRemove}, {@link UpdateMerge}
 */
export type UpdateOp =
  | UpdateSet
  | UpdateIncrement
  | UpdateDecrement
  | UpdateAppend
  | UpdateRemove
  | UpdateMerge

/** Input for atomically updating a stream item. */
export type StreamUpdateInput = {
  /** Name of the stream. */
  stream_name: string
  /** Group identifier. */
  group_id: string
  /** Item identifier. */
  item_id: string
  /** Ordered list of update operations to apply atomically. */
  ops: UpdateOp[]
}

/** Trigger config for `stream` triggers. Filters which item changes fire the handler. */
export interface StreamTriggerConfig {
  /** Stream name to watch. Only changes on this stream fire the handler. */
  stream_name: string
  /** If set, only changes within this group fire the handler. */
  group_id?: string
  /** If set, only changes to this specific item fire the handler. */
  item_id?: string
  /** Function ID for conditional execution. If it returns `false`, the handler is skipped. */
  condition_function_id?: string
}

/** Trigger config for `stream:join` and `stream:leave` triggers. */
export interface StreamJoinLeaveTriggerConfig {
  /** Function ID for conditional execution. If it returns `false`, the handler is skipped. */
  condition_function_id?: string
}

/** Handler input for `stream` triggers, fired when an item changes via `stream::set`, `stream::update`, or `stream::delete`. */
export interface StreamChangeEvent {
  /** The event type. */
  type: 'stream'
  /** Unix timestamp of the event. */
  timestamp: number
  /** The stream where the change occurred. */
  streamName: string
  /** The group where the change occurred. */
  groupId: string
  /** The item ID that changed. */
  id?: string
  /** The event detail object containing `type` and `data` fields. */
  event: {
    type: 'create' | 'update' | 'delete'
    // biome-ignore lint/suspicious/noExplicitAny: any is fine here
    data: any
  }
}

/**
 * Interface for custom stream implementations. Passed to `ISdk.createStream`
 * to override the engine's built-in stream storage.
 *
 * @typeParam TData - Type of the data stored in the stream.
 */
export interface IStream<TData> {
  /** Retrieve a single item by group and item ID. */
  get(input: StreamGetInput): Promise<TData | null>
  /** Set (create or overwrite) a stream item. */
  set(input: StreamSetInput): Promise<StreamSetResult<TData> | null>
  /** Delete a stream item. */
  delete(input: StreamDeleteInput): Promise<DeleteResult>
  /** List all items in a group. */
  list(input: StreamListInput): Promise<TData[]>
  /** List all group IDs in a stream. */
  listGroups(input: StreamListGroupsInput): Promise<string[]>
  /** Apply atomic update operations to a stream item. */
  update(input: StreamUpdateInput): Promise<StreamUpdateResult<TData> | null>
}
